wpilib/wpilib/spi.py

```diff
@@ -198,103 +201,331 @@ def read(self, initiate, size):
+
+        :meth:`.initAuto` and :meth:`.setAutoTransmitData` must
+        be called before calling this function.
+
+        :param period: period between transfers, in seconds (us resolution)
+        """
+        hal.startSPIAutoRate(self.port, period)
+
+    def startAutoTrigger(self, source: DigitalSource, rising: bool, falling: bool) -> None:
+        """Start running the automatic SPI transfer engine when a trigger occurs.
+
+        :meth:`.initAuto` and :meth:`.setAutoTransmitData` must
+        be called before calling this function.
+
+        :param source: digital source for the trigger (may be an analog trigger)
+        :param rising: trigger on the rising edge
+        :param falling: trigger on the falling edge
+        """
+        hal.startSPIAutoTrigger(self.port, source.getPortHandleForRouting(),
+                                source.getAnalogTriggerTypeForRouting(),
+                                rising, falling)
+
+    def stopAuto(self) -> None:
+        """Stop running the automatic SPI transfer engine."""
+        hal.stopSPIAuto(self.port)
+
+    def forceAutoRead(self) -> None:
+        """Force the engine to make a single transfer."""
+        hal.forceSPIAutoRead(self.port)
+
+    def readAutoReceivedData(self, buffer, numToRead: int, timeout: float) -> (int, bytes):
+        """Read data that has been transferred by the automatic SPI transfer engine.
+
+        Transfers may be made a byte at a time, so it's necessary for the caller
+        to handle cases where an entire transfer has not been completed.
+
+        Blocks until numToRead bytes have been read or timeout expires.
+        May be called with numToRead=0 to retrieve how many bytes are available.
+
+        :param buffer: A ctypes c_uint8 buffer to read the data into
+        :param numToRead: number of bytes to read
+        :param timeout: timeout in seconds (ms resolution)
+        :returns: Number of bytes remaining to be read
+        """
+        if len(buffer) < numToRead:
+            raise ValueError("buffer is too small, must be at least %s" % numToRead)
+        return hal.readSPIAutoReceivedData(self.port, buffer, numToRead, timeout)
+
+    def getAutoDroppedCount(self) -> int:
+        """Get the number of bytes dropped by the automatic SPI transfer engine due
+        to the receive buffer being full.
+
+        :returns: Number of bytes dropped
+        """
+        return hal.getSPIAutoDroppedCount(self.port)
+
+    class _Accumulator:
+
+        kAccumulateDepth = 2048
+
+        def __init__(self, port: int, xferSize: int, validMask: int,
+                           validValue: int, dataShift: int, dataSize: int,
+                           isSigned: bool, bigEndian: bool):
+
+            # python-specific: for efficiency purposes, we only support xferSize of
+            #                  2, 4, or 8... if you need to do a different size then
+            #                  file a bug and we'll adjust it
```

Comment:
Hm. Is using `struct` actually any faster than `int.from_bytes`? I mean I wouldn't be surprised if it is, but I guess numbers would be nice.

Reply:
well, from_bytes doesn't support a buffer offset? I think.

Reply:
Ah, good point; I missed that part. You could definitely use a `memoryview` to do that though.